activerecord/test/cases/nested_attributes_test.rb

```diff
@@ -1117,3 +1119,100 @@ def test_extend_affects_nested_attributes
     assert_equal "from extension", pirate.treasures[0].name
   end
 end
+
+class TestHasManyNestedAttributesWithPrimaryKey < ActiveRecord::TestCase
+  setup do
+    @author = Author.create!(name: "Blackbeard")
+    @book = @author.books.create!(name: "Midsummer Night's Pillage", isbn: "12345")
+  end
+
+  test "when no :primary_key is provided, defaults to association's primary key" do
+    Author.accepts_nested_attributes_for :books
+
+    @author.update(books_attributes: [{ id: @book.id, name: "Pillaging for Dummies" }])
+
+    assert_equal "Pillaging for Dummies", @book.reload.name
+  end
+
+  test "when :primary_key is provided for an existing record, the record is updated" do
+    Author.accepts_nested_attributes_for :books, primary_key: :isbn
+
+    @author.update(books_attributes: [{ id: "12345", name: "Pillaging for Dummies" }])
```

Comment:
@nickrivadeneira `isbn` instead of `id` here?

Reply:
@akshaymohite It's `id` by design to mirror the behavior that you'd see if you were to set `self.primary_key = :isbn` on the `Book` model. In this case, the current behavior would be that the client sends `{ id: "12345", ... }` and Rails would query based on `Book#id`, which resolves to the configured primary key, or `:isbn` in our scenario.

I mirrored the existing behavior because I think it's a good idea to abstract away the implementation details of what column you're using as an identifier, and allow the client to speak in terms of "Here is the identifier of the record I'm interested in, regardless of how you store that".